File: dawn_haskell/src/Tango.hs
module AuxTree (Tree, empty, insert, find) where

data Tree a = Empty | Node (Tree a) a (Tree a)
    deriving (Show)

empty :: Tree a
empty = Empty

insert :: Ord a => a -> Tree a -> Tree a
insert x Empty = Node Empty x Empty
insert x (Node left y right)
    | x < y     = Node (insert x left) y right
    | otherwise = Node left y (insert x right)

find :: Ord a => a -> Tree a -> Bool
find x Empty = False
find x (Node left y right)
    | x == y    = True
    | x < y     = find x left
    | otherwise = find x right




module TangoTree (TangoTree, empty, find, update) where

import AuxTree (Tree, empty, insert, find)

type TangoTree a = [Tree a]

empty :: TangoTree a
empty = [empty]

find :: Ord a => a -> TangoTree a -> Bool
find x = any (find x) 

split :: Ord a => a -> Tree a -> (Tree a, Tree a)
split x Empty = (Empty, Empty)
split x (Node left y right)
    | x < y     = (left1, Node left2 y right)
    | otherwise = (Node left y right1, right2)
    where
        (left1, left2) = split x left
        (right1, right2) = split x right

join :: Tree a -> Tree a -> Tree a
join Empty right = right
join left Empty = left
join left@(Node ll x lr) right@(Node rl y rr)
    | x < y     = Node ll x (join lr right)
    | otherwise = Node (join left rl) y rr

update :: Ord a => a -> TangoTree a -> TangoTree a
update x [] = [Node Empty x Empty]
update x (aux:rest) = join left (Node Empty x Empty) : rest
    where
        (left, right) = split x aux

File: dawn_ocaml/lib/tango.ml
open Base

module AuxTree = struct
  type key = int
  type t = 
    | Empty
    | Node of t * key * t

  let empty = Empty

  let rec insert x t =
    match t with
    | Empty -> Node (Empty, x, Empty)
    | Node (l, y, r) ->
       if x < y then Node (insert x l, y, r)
       else Node (l, y, insert x r)

  let rec find x t =
    match t with
    | Empty -> false
    | Node (l, y, r) ->
       if x = y then true
       else if x < y then find x l
       else find x r
end

module TangoTree = struct
  type key = int
  type t = AuxTree.t list

  let empty = [AuxTree.empty]

  let rec find x t =
    match t with
    | [] -> false
    | aux :: rest ->
       if AuxTree.find x aux then true
       else find x rest

  let rec split x aux =
    match aux with
    | AuxTree.Empty -> (AuxTree.empty, AuxTree.empty)
    | AuxTree.Node (l, y, r) ->
       if x < y then
         let (ll, lr) = split x l in
         (ll, AuxTree.Node (lr, y, r))
       else
         let (rl, rr) = split x r in
         (AuxTree.Node (l, y, rl), rr)

  let rec join l r =
    match (l, r) with
    | (AuxTree.Empty, _) -> r
    | (_, AuxTree.Empty) -> l
    | (AuxTree.Node (ll, lx, lr), AuxTree.Node (rl, rx, rr)) ->
       if lx < rx then AuxTree.Node (ll, lx, join lr r)
       else AuxTree.Node (join l rl, rx, rr)

  let rec update x t =
    match t with
    | [] -> [AuxTree.Node (AuxTree.empty, x, AuxTree.empty)]
    | aux :: rest ->
       let (l, r) = split x aux in
       join l (AuxTree.Node (AuxTree.empty, x, AuxTree.empty)) :: rest
end